test/unit/app/validators/relative_date_validator_test.rb

```diff
@@ -0,0 +1,66 @@
+
+class RelativeDateValidatorTest < ActiveSupport::TestCase
+  class StubModel
+    include ActiveModel::API
+    attr_accessor :some_date
+  end
+
+  test "validation is not applied if attribute value is nil" do
+    model = StubModel.new(some_date: nil)
+    validator = RelativeDateValidator.new(attributes: :some_date, before: -> { Time.zone.local(2022, 1, 1, 1, 3) }, after: -> { Time.zone.local(2022, 1, 1, 1, 1) })
+    validator.validate(model)
+    assert_empty model.errors
+  end
+
+  test "model is valid if datetime is between before and after dates" do
+    model = StubModel.new(some_date: Time.zone.local(2022, 1, 1, 1, 2))
+    validator = RelativeDateValidator.new(attributes: :some_date, before: -> { Time.zone.local(2022, 1, 1, 1, 3) }, after: -> { Time.zone.local(2022, 1, 1, 1, 1) })
+    validator.validate(model)
+    assert_empty model.errors
+  end
+
+  test "adds an error if date is later than before date" do
+    model = StubModel.new(some_date: Time.zone.local(2022, 1, 1, 1, 2))
+    validator = RelativeDateValidator.new(attributes: :some_date, before: -> { Time.zone.local(2022, 1, 1, 1, 1) })
+    validator.validate(model)
+    assert_not_empty model.errors
+  end
+
+  test "adds an error if date is equal to before date" do
+    model = StubModel.new(some_date: Time.zone.local(2022, 1, 1, 1, 1))
+    validator = RelativeDateValidator.new(attributes: :some_date, before: -> { Time.zone.local(2022, 1, 1, 1, 1) })
+    validator.validate(model)
+    assert_not_empty model.errors
+  end
+
+  test "adds an error if date is earlier than after date" do
+    model = StubModel.new(some_date: Time.zone.local(2022, 1, 1, 1, 1))
+    validator = RelativeDateValidator.new(attributes: :some_date, after: -> { Time.zone.local(2022, 1, 1, 1, 2) })
+    validator.validate(model)
+    assert_not_empty model.errors
+  end
+
+  test "adds an error if date is equal to after date" do
+    model = StubModel.new(some_date: Time.zone.local(2022, 1, 1, 1, 1))
+    validator = RelativeDateValidator.new(attributes: :some_date, after: -> { Time.zone.local(2022, 1, 1, 1, 1) })
+    validator.validate(model)
+    assert_not_empty model.errors
+  end
+
+  test "before error message can be overriden" do
+    before_message = "should be in the past"
+    model = StubModel.new(some_date: Time.zone.local(2022, 1, 1, 1, 2))
+    validator = RelativeDateValidator.new(attributes: :some_date, before: -> { Time.zone.local(2022, 1, 1, 1, 1) }, before_message:)
+    validator.validate(model)
+    assert_equal before_message, model.errors[:some_date].first
+  end
+
+  test "after error message can be overriden" do
+    after_message = "should be in the future"
+    model = StubModel.new(some_date: Time.zone.local(2022, 1, 1, 1, 1))
+    validator = RelativeDateValidator.new(attributes: :some_date, after: -> { Time.zone.local(2022, 1, 1, 1, 2) }, after_message:)
+    validator.validate(model)
+    assert_equal after_message, model.errors[:some_date].first
+  end
```

Comment:
It might be nice to have tests to cover the _default_ message too. The other tests only `assert_not_empty model.errors` rather than the message itself.

Reply:
Good point, there's a bug in there because I'm not calling the lambda before attempting to cast it to a string 😱